File: killer_sudoku/solver/engine/rule.py
"""SolverRule protocol, RuleContext, and RuleStats.

SolverRule is a structural protocol — any object with the required
attributes and an apply() method qualifies. Rules are stateless;
all mutable state lives in BoardState. Rules must not call BoardState
mutators directly; they return RuleResult.
"""

from __future__ import annotations

import dataclasses
from typing import Protocol

from killer_sudoku.solver.engine.board_state import BoardState
from killer_sudoku.solver.engine.hint import HintResult
from killer_sudoku.solver.engine.types import (
    Cell,
    Elimination,
    RuleResult,
    Trigger,
    Unit,
    UnitKind,
)


@dataclasses.dataclass
class RuleContext:
    """Input to a rule's apply() method."""

    unit: Unit | None  # None for CELL_DETERMINED and GLOBAL rules
    cell: Cell | None  # Set for CELL_DETERMINED; None otherwise
    board: BoardState
    hint: Trigger
    hint_digit: int | None  # Digit that triggered, if known


@dataclasses.dataclass
class RuleStats:
    """Accumulated statistics for a single rule across all solves."""

    calls: int = 0
    progress: int = 0  # times apply() produced any result
    eliminations: int = 0  # total candidate Eliminations returned
    elapsed_ns: int = 0

    @property
    def hit_rate(self) -> float:
        """Fraction of apply() calls that made progress."""
        return self.progress / self.calls if self.calls else 0.0

    @property
    def utility(self) -> float:
        """Eliminations per nanosecond; used for offline priority calibration."""
        cost = self.elapsed_ns / self.calls if self.calls else 1.0
        return (self.eliminations / self.calls if self.calls else 0.0) / cost

    def record(self, result: RuleResult, elapsed_ns: int) -> None:
        """Record one apply() call's results."""
        self.calls += 1
        if result.has_progress:
            self.progress += 1
        self.eliminations += len(result.eliminations)
        self.elapsed_ns += elapsed_ns


class SolverRule(Protocol):
    """Structural protocol for solver rules.

    Rules are stateless — apply() reads from ctx.board and returns a
    RuleResult.  It must not call any BoardState mutator directly.
    unit_kinds: empty frozenset means GLOBAL/cell-scoped (unit=None in ctx).
    """

    name: str
    description: str
    priority: int
    triggers: frozenset[Trigger]
    unit_kinds: frozenset[UnitKind]

    def apply(self, ctx: RuleContext) -> RuleResult:
        """Apply the rule; return eliminations, placements, solution prunings, or
        virtual cage additions.  Rules that only produce candidate eliminations
        construct RuleResult(eliminations=[...]).
        """
        ...

    def as_hints(
        self, ctx: RuleContext, eliminations: list[Elimination]
    ) -> list[HintResult]:
        """Convert a rule firing to coaching hints (called when rule is hint-only).

        ctx provides the triggering context (unit, cell, board, trigger, hint_digit).
        eliminations is result.eliminations from apply() for the same ctx.
        Rules that produce placement hints (e.g. NakedSingle) may ignore eliminations.
        Rules that produce multiple independent hints (e.g. CageConfinement) return
        one HintResult per finding.
        """
        ...

File: web/src/engine/rule.ts
/**
 * SolverRule interface, RuleContext, and RuleStats.
 *
 * SolverRule is a structural interface — any object with the required
 * fields and an apply() method qualifies. Rules are stateless; all mutable
 * state lives in KillerBoardState.
 */

import { KillerBoardState } from './boardState.js';
import type { BoardState } from './boardState.js';
import type { HintResult } from './hint.js';
import { emptyResult } from './types.js';
import type { Cell, Elimination, RuleResult, Trigger, Unit, UnitKind } from './types.js';

/** Input to a rule's apply() method. */
export interface RuleContext {
  /** null for CELL_DETERMINED and GLOBAL rules. */
  readonly unit: Unit | null;
  /** Set for CELL_DETERMINED; null otherwise. */
  readonly cell: Cell | null;
  readonly board: BoardState;
  readonly hint: Trigger;
  readonly hintDigit: number | null;
}

/** Accumulated statistics for a single rule across all solves. */
export interface RuleStats {
  calls: number;
  progress: number;
  eliminations: number;
  elapsedNs: number;
}

/** Convenience constructor: zeroed RuleStats for a newly registered rule. */
export function makeRuleStats(): RuleStats {
  return { calls: 0, progress: 0, eliminations: 0, elapsedNs: 0 };
}

/** Structural interface for solver rules. */
export interface SolverRule {
  readonly name: string;
  readonly displayName: string;
  readonly description: string;
  readonly priority: number;
  /** True when this rule requires killer cage constraints and must be excluded for classic puzzles. */
  readonly killerOnly: boolean;
  readonly triggers: ReadonlySet<Trigger>;
  /**
   * Empty set means GLOBAL / cell-scoped (unit=null in ctx).
   * For unit-scoped rules, the set lists which UnitKind values apply.
   */
  readonly unitKinds: ReadonlySet<UnitKind>;

  apply(ctx: RuleContext): RuleResult;

  asHints(ctx: RuleContext, eliminations: readonly Elimination[]): HintResult[];
}

/** Narrows `RuleContext.board` to `KillerBoardState` for killer-only rules —
 *  `ctx.board.linearSystem`/`regions`/`cageSolns`/`spec` are then directly
 *  typed, with no cast and no further narrow inside the rule body. */
export interface KillerRuleContext extends Omit<RuleContext, 'board'> {
  readonly board: KillerBoardState;
}

/**
 * Shared base for the ten rules that require killer cage constraints
 * (`deltaConstraint`, `linearElimination`, `sumPairConstraint`,
 * `cageCandidateFilter`, `cageConfinement`, `cageIntersection`, `mustContain`,
 * `mustContainOutie`, `solutionMapFilter`, `unitPartitionFilter`).
 *
 * Performs the `ctx.board instanceof KillerBoardState` narrow exactly once —
 * in `apply`/`asHints` — and hands subclasses a `KillerRuleContext` whose
 * `board` is directly typed as `KillerBoardState`. `killerOnly` is set here so
 * it is never repeated across the ten subclasses.
 *
 * `buildEngine` (via `PuzzleState.isKiller`) already filters `killerOnly`
 * rules out of the classic rule set, so the `instanceof` branch below is
 * unreachable in practice — but the type system still requires *some* narrow
 * to expose `KillerBoardState`'s members to `applyKiller`/`asHintsKiller`,
 * and this is the one place it lives (defense in depth).
 */
export abstract class KillerOnlyRule implements SolverRule {
  readonly killerOnly = true;

  apply(ctx: RuleContext): RuleResult {
    if (!(ctx.board instanceof KillerBoardState)) return emptyResult();
    return this.applyKiller({ ...ctx, board: ctx.board });
  }

  asHints(ctx: RuleContext, eliminations: readonly Elimination[]): HintResult[] {
    if (!(ctx.board instanceof KillerBoardState)) return [];
    return this.asHintsKiller({ ...ctx, board: ctx.board }, eliminations);
  }

  abstract readonly name: string;
  abstract readonly displayName: string;
  abstract readonly description: string;
  abstract readonly priority: number;
  abstract readonly triggers: ReadonlySet<Trigger>;
  abstract readonly unitKinds: ReadonlySet<UnitKind>;
  abstract applyKiller(ctx: KillerRuleContext): RuleResult;
  abstract asHintsKiller(ctx: KillerRuleContext, eliminations: readonly Elimination[]): HintResult[];
}
